fetcher/fetcher_test.go

```diff
@@ -0,0 +1,123 @@
+package fetcher
+
+import (
+	"context"
+	"fmt"
+	"github.com/algorand/go-algorand-sdk/client/v2/algod"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/rpcs"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+)
+
+type AlgodHandler struct {
+	mock.Mock
+}
+
+func (handler *AlgodHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	handler.Called(w, req)
+	return
+}
+
+type BlockHandler struct {
+	mock.Mock
+}
+
+func (handler *BlockHandler) handlerFunc(ctx context.Context, cert *rpcs.EncodedBlockCert) error {
+	args := handler.Called(ctx, cert)
+	return args.Error(0)
+}
+
+func mockAClient(t *testing.T, algodHandler *AlgodHandler) *algod.Client {
+	mockServer := httptest.NewServer(algodHandler)
+	aclient, err := algod.MakeClient(mockServer.URL, "")
+	if err != nil {
+		t.FailNow()
+	}
+	return aclient
+}
+
+func TestFetcherImplErrorInitialization(t *testing.T) {
+	aclient := mockAClient(t, &AlgodHandler{})
+	fetcher := &fetcherImpl{aclient: aclient, log: logrus.New()}
+	require.Equal(t, "", fetcher.Error(), "Initialization of fetcher caused an unexpected error.")
+}
+
+func TestFetcherImplAlgodReturnsClient(t *testing.T) {
+	aclient := mockAClient(t, &AlgodHandler{})
+	fetcher := &fetcherImpl{aclient: aclient, log: logrus.New()}
+	require.Equal(t, aclient, fetcher.Algod(), "Algod client returned from fetcherImpl does not match expected instance.")
+}
+
+func TestFetcherImplSetError(t *testing.T) {
+	aclient := mockAClient(t, &AlgodHandler{})
+	fetcher := &fetcherImpl{aclient: aclient, log: logrus.New()}
+	expectedErr := fmt.Errorf("foobar")
+	fetcher.setError(expectedErr)
+	require.Equal(t, expectedErr.Error(), fetcher.Error(), "Error produced by setError was not reflected in Error output.")
+}
+
+func TestFetcherImplProcessQueueHandlerError(t *testing.T) {
+	mockAlgodHandler := &AlgodHandler{}
+	aclient := mockAClient(t, mockAlgodHandler)
+	ctx, cf := context.WithCancel(context.Background())
+	defer cf()
```

Comment:
I guess this is to ensure fetcher terminates after the test is finished?

If so, I don't see how it could work. If there is no error, wouldn't `fetcher.Run` go on forever?

Perhaps `context.WithTimeout` would be a better safeguard?

What happens if you don't use either, and just call `fetcher.Run(context.Background())`?

Reply:
You're right, these are completely unnecessary. I removed these and just used `context.Background()` for the tests.